analysis/topology/sequences.lean

```diff
@@ -0,0 +1,278 @@
+  converges_to x limit ↔ tendsto x at_top (nhds limit) := 
+  iff.intro
+    (assume xtol : converges_to x limit,
+      suffices ∀ U, limit ∈ U → is_open U → x ⁻¹' U ∈ at_top.sets, from tendsto_nhds this,
+        assume U limitInU isOpenU,
+          suffices ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, by simp[this],
+          xtol U limitInU isOpenU)
+    (assume ttol : tendsto x at_top (nhds limit),
+      show ∀ U : set X, limit ∈ U → is_open U → ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, from
+        assume U limitInU isOpenU,
+          have {n | (x n) ∈ U} ∈ at_top.sets, from mem_map.mp $ le_def.mp ttol U 
+                                               (mem_nhds_sets isOpenU limitInU),
+          show ∃ n0 : ℕ, ∀ n ≥ n0, (x n) ∈ U, from mem_at_top_sets.mp this)
+
+/- The sequential closure of a subset M ⊆ X of a topological space X is 
+ - the set of all p ∈ X which arise as limit of sequences in M.
+ -/
+def sequential_closure (M : set X) : set X :=
+{p | ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ M)) ∧ (converges_to x p)}
+
+
+lemma subset_seq_closure (M : set X) : M ⊆ sequential_closure M := 
+assume p (_ : p ∈ M), show p ∈ sequential_closure M, from
+  exists.intro (λ n, p) ⟨assume n, ‹p ∈ M›, const_seq_conv p⟩
+
+def is_seq_closed (A : set X) : Prop := A = sequential_closure A
+
+/- A convenience lemma for showing that a set is sequentially closed. -/
+lemma is_seq_closed_of_def {A : set X} (h : ∀ (x : ℕ → X), (∀ n : ℕ, ((x n) ∈ A)) → ∀ p : X,
+  converges_to x p → p ∈ A) : is_seq_closed A :=
+show A = sequential_closure A, from set.ext (assume p, iff.intro
+  (assume : p ∈ A, subset_seq_closure A ‹p ∈ A›)
+  (assume : p ∈ sequential_closure A, 
+    have ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ A)) ∧ (converges_to x p), by assumption,
+    let ⟨x, ⟨_, _⟩⟩ := this in
+    show p ∈ A, from h x ‹∀ n : ℕ, ((x n) ∈ A)› p ‹converges_to x p›))
+
+/- The sequential closure of a set is contained in the closure of that set. The converse is not
+ - true.
+ -/ 
+lemma sequential_closure_subset_closure (M : set X) : sequential_closure M ⊆ closure M :=
+show ∀ p, p ∈ sequential_closure M → p ∈ closure M, from
+assume p,
+assume : ∃ x : ℕ → X, (∀ n : ℕ, ((x n) ∈ M)) ∧ (converges_to x p),
+let ⟨x, ⟨_, _⟩⟩ := this in
+show p ∈ closure M, from
+-- we have to show that p is in the closure of M
+-- using mem_closure_iff, this is equivalent to proving that every open neighbourhood
+-- has nonempty intersection with A, but this is witnessed by our sequence x
+suffices ∀ O, is_open O → p ∈ O → O ∩ M ≠ ∅, from mem_closure_iff.mpr this,
+assume O is_open_O O_cap_M_neq_empty,
+let ⟨n0, _⟩ := ‹converges_to x p› O ‹p ∈ O› ‹is_open O› in
+have (x n0) ∈ O, from ‹∀ n ≥ n0, x n ∈ O› n0 (show n0 ≥ n0, from le_refl n0),
+have (x n0) ∈ O ∩ M, from ⟨this, ‹∀n, x n ∈ M› n0⟩,
+set.ne_empty_of_mem this
+
+/- A set is sequentially closed if it is closed. -/
+lemma is_seq_closed_of_is_closed : ∀ M : set X, is_closed M → is_seq_closed M :=
```

Comment:
Normally we move hypotheses left of the colon when it makes the proof simpler, like
```suggestion
lemma is_seq_closed_of_is_closed (M : set X) (h : is_closed M) : is_seq_closed M :=
```
and then delete the next line.